servers/republik/modules/crowdfundings/lib/generateMemberships.js

```diff
@@ -60,52 +62,123 @@ module.exports = async (pledgeId, pgdb, t, req, logger = console) => {
   let cancelableMemberships = []
   let membershipPeriod
   const now = new Date()
-  pledgeOptions.forEach((plo) => {
+
+  await Promise.map(pledgeOptions, async (plo) => {
     if (plo.packageOption.reward.type === 'MembershipType') {
-      const membershipType = plo.packageOption.reward.membershipType
-      for (let c = 0; c < plo.amount; c++) {
-        const membership = {
-          userId: user.id,
-          pledgeId: pledge.id,
-          membershipTypeId: membershipType.id,
-          reducedPrice,
-          voucherable: !reducedPrice,
-          active: false,
-          renew: false,
-          createdAt: now,
-          updatedAt: now
-        }
+      // Is amount in pledgeOption > 0?
+      if (plo.amount === 0) {
+        debug('peldgeOption amount is 0')
+        return
+      }
 
-        if (
-          c === 0 &&
-          !membershipPeriod &&
-          !userHasActiveMembership &&
-          pkg.isAutoActivateUserMembership
-        ) {
-          membershipPeriod = {
-            beginDate: now,
-            endDate: moment(now).add(membershipType.intervalCount, membershipType.interval),
-            membership
+      const { membershipId, autoPay = false } = plo.customization
+      const { membershipType } = plo.packageOption.reward
+
+      if (membershipId) {
+        debug('membershipId "%s"', membershipId)
+
+        const membership =
+          await pgdb.public.memberships.findOne({ id: membershipId })
+
+        const membershipPeriods =
+          await pgdb.public.membershipPeriods.find({ membershipId })
+
+        const membershipUser =
+          await pgdb.public.users.findOne({ id: membership.userId })
+
+        Object.assign(
+          membership,
+          {
+            membershipType,
+            membershipPeriods,
+            user: membershipUser
           }
-        } else {
-          // Cancel active memberships because bought package (option) contains
-          // a better abo.
-          if (['ABO', 'BENEFACTOR_ABO'].includes(membershipType.name)) {
-            cancelableMemberships =
-              activeMemberships
-                .filter(m => (m.name === 'MONTHLY_ABO' && m.renew === true))
+        )
+
+        const resolvedPackage = (await resolvePackages({
+          packages: [pkg],
+          pledger: user,
+          pgdb
+        })).shift()
+
+        const { additionalPeriods } =
+          await evaluate({
+            package_: resolvedPackage,
+            packageOption: plo.packageOption,
+            membership
+          })
+
+        await pgdb.public.membershipPeriods.insert(
+          additionalPeriods
+            .map(period => omit(period, ['id', 'createdAt', 'updatedAt']))
+        )
+
+        await pgdb.public.memberships.update(
+          { id: membershipId },
+          {
+            autoPay,
+            active: true,
+            renew: true,
+            updatedAt: now
```

Comment:
why is the `membershipTypeId` not updated here. If I upgrade on prolong I want to have the approriate type from then on, or how is this achived?

Reply:
Not inclined to alter `membership.membershipTypeId` as it would alter data into the past with current set of queries.

Suggesting this: `membership.membershipTypeId` becomes a mere reflection of current state of membership and `membershipPeriod.membershipTypeId` sort of keeps track.

Or `membership` describes initial state, latest `membershipPeriod` reflects how it was altered with `membershipTypeId` and `pledgeId` (which caused period). Prefer this one.

Reply:
I also prefer the later one. I see you already added the `membershipPeriod` <-> `pledgeOption` relationship. I suggest we resolve the (current) type in the [Membership resolver](https://github.com/orbiting/backends/blob/master/servers/republik/modules/crowdfundings/graphql/resolvers/Membership.js#L2) by checking the `membershipTypeId` of the latest period. So it would always be resolved to the current type, but we don't lose information about what it initially was.
We could also add `Membership.initialType` resolver that would do what the current one does. But that's optional now.